api/src/main/java/org/openmrs/module/pihcore/event/PatientUpdateEventConsumer.java

```diff
@@ -0,0 +1,212 @@
+    private Rocks statusDb;
+    private boolean snapshotInitialized = false;
+
+    public PatientUpdateEventConsumer(DbEventSourceConfig config) {
+        this.config = config;
+        this.database = config.getContext().getDatabase();
+        patientKeys.put("person_id", "person");
+        patientKeys.put("person_a", "person");
+        patientKeys.put("person_b", "person");
+        patientKeys.put("order_id", "orders");
+        patientKeys.put("patient_program_id", "patient_program");
+        patientKeys.put("encounter_id", "encounter");
+        patientKeys.put("visit_id", "visit");
+        patientKeys.put("allergy_id", "allergy");
+        patientKeys.put("appointment_id", "appointmentscheduling_appointment");
+    }
+
+    @Override
+    public void startup() {
+        try {
+            statusDb = new Rocks(new File(config.getContext().getModuleDataDir(), "status.db"));
+            snapshotInitialized = BooleanUtils.isTrue(statusDb.get("snapshotInitialized"));
+        }
+        catch (Exception e) {
+            throw new RuntimeException("An error occurred starting up", e);
+        }
+    }
+
+    @Override
+    public void shutdown() {
+        statusDb.close();
+    }
+
+    @Override
+    public void accept(DbEvent event) {
+        if (!snapshotInitialized) {
+            performInitialSnapshot();
+            statusDb.put("snapshotInitialized", Boolean.TRUE);
+            snapshotInitialized = true;
+        }
+        Set<Integer> patientIds = getPatientIdsForEvent(event);
+        if (patientIds.isEmpty()) {
+            log.warn("Not handling event as not mapped to patient: " + event);
+        }
+        else {
+            for (Integer patientId : patientIds) {
+                handlePatientEvent(patientId, event);
+            }
+        }
+    }
+
+    /**
+     * For the given event, returns the associated patient ids
+     * @param event the event to retrieve the related patient ids
+     * @return the patient ids associated with the given event
+     */
+    public Set<Integer> getPatientIdsForEvent(DbEvent event) {
+        Set<Integer> ret = new HashSet<>();
+        Integer patientId = event.getValues().getInteger("patient_id");
+        if (patientId != null) {
+            ret.add(patientId);
+        }
+        else {
+            for (String key : patientKeys.keySet()) {
+                String keyTable = patientKeys.get(key);
+                Integer value = event.getValues().getInteger(key);
+                if (value != null) {
+                    SqlBuilder sql = new SqlBuilder();
+                    sql.select("p.patient_id").from("patient", "p");
+                    if (keyTable.equals("person")) {
+                        sql.where("p.patient_id = ?");
+                    } else {
+                        sql.innerJoin(keyTable, "x", "patient_id", "p", "patient_id");
+                        sql.where("x." + key + " = ?");
```

Comment:
So, for example, there might be a change to the encounter table, but if the patient_id doesn't change (which likely never will!) it won't be in the event, so we fetch the patient id this way? Makes sense.

Reply:
Not exactly.  The patient_id is always in the event (the event contains the full before and after values, not just the diffs).  This piece of code wouldn't ever get hit for the "encounter" table, as an encounter event has a patient_id directly, so it would just get used.  This piece of code would get hit for something like "encounter_provider", which is patient data but only links to the patient via the encounter table.  So we need to first join encounter_provider to encounter in order to retrieve the associated patient_id from encounter.

Reply:
aha, that makes sense, and explains my confusion regarding obs as well, thanks....